datafusion/src/physical_optimizer/pruning.rs

```diff
@@ -28,50 +28,75 @@
 //! https://github.com/apache/arrow-datafusion/issues/363 it will
 //! be genericized.
 
-use std::{collections::HashSet, sync::Arc};
+use std::{collections::HashSet, convert::TryInto, sync::Arc};
 
 use arrow::{
-    array::{
-        make_array, new_null_array, ArrayData, ArrayRef, BooleanArray,
-        BooleanBufferBuilder,
-    },
-    buffer::MutableBuffer,
-    datatypes::{DataType, Field, Schema},
+    array::{ArrayRef, BooleanArray},
+    datatypes::{Field, Schema, SchemaRef},
     record_batch::RecordBatch,
 };
 
-use parquet::file::{
-    metadata::RowGroupMetaData, statistics::Statistics as ParquetStatistics,
-};
-
 use crate::{
     error::{DataFusionError, Result},
     execution::context::ExecutionContextState,
     logical_plan::{Expr, Operator},
     optimizer::utils,
     physical_plan::{planner::DefaultPhysicalPlanner, ColumnarValue, PhysicalExpr},
+    scalar::ScalarValue,
 };
 
+/// Interface to pass statistics information to [`PruningPredicates`]
+pub trait PruningStatistics {
+    /// return the minimum value for the named column, if known
+    fn min_value(&self, column: &str) -> Option<ScalarValue>;
```

Comment:
One concern I have (in general with current state of DataFusion) is that we use `ScalarValue` a lot in code which can detrimental to performance in some cases (compared to a typed array).
Also in this case, if we would have a large dataset with 1000s of statistics values, calculating statistics might be slower than it could be when it is stored in contiguous memory.
Just a thought - not something we should block this PR for.

Reply:
what would be a better alternative in the long run? generic?

Reply:
In the longer run I would say we should start pushing more towards typed contiguous arrays (`Array`s or `Vec`), indeed generic. For example, here the min and max values per group could be stored in two arrays of corresponding types which would be faster and uses less memory.

Vectorized processing is what we try to use Arrow for already, so I would say it is good to try to use it in more places where it makes sense.

Reply:
I very much agree with you, @Dandandan . I have a PR on arrow2 exactly on this on arrow2. https://github.com/jorgecarleitao/arrow2/pull/56

I am waiting for the experimental repos to be available so that we can discuss it further in apache/*

Reply:
Ok, I can take a shot at trying to do this.

@Dandandan  are you thinking something like:

```
pub trait PruningStatistics {
    fn min_values(&self, column: &str) -> Option<ArrayRef>;
    fn max_values(&self, column: &str) -> Option<ArrayRef>;
} 
```

Which would be constrained to return Arrays the same number of rows?

Reply:
@alamb yes - something along those lines.

Reply:
I'll give it a go and see what it looks like

Reply:
Here is what the alternate API looks like https://github.com/apache/arrow-datafusion/pull/426

I would say it is now harder to implement the `PruningStatistics` API in the parquet reader, but the implementation in `pruning.rs` is simpler